### PR TITLE
CASMCMS-5793: Add v3 API

### PR DESCRIPTION
**Pull request by rbak-hpe**
## Summary and Scope

Adds a new v3 api with support for paging, additional debugging features and links to the ARA records

## Issues and Related PRs

* Resolves CASMCMS-5793
* Resolves CASMCMS-7707
* Resolves CASMCMS-8625

### Tested on:

  * Mug/Dorian/Beau

### Test description:

Tested every endpoint for v2 and v3 as well as manual and automatic session creation

## Risks and Mitigations

None

## Pull Request Checklist

- [X] Version number(s) incremented, if applicable
- [X] Copyrights updated
- [X] License file intact
- [X] Target branch correct
- [X] CHANGELOG.md updated
- [X] Testing is appropriate and complete, if applicable
- [X] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

**Comment by mharding-hpe** [comment]
I rebased this to resolve merge conflicts

**Review by jsl-hpe — APPROVED** [review]

File: src/server/cray/cfs/api/controllers/components.py [hunk 1]
```diff
@@ -218,41 +297,156 @@ def patch_v2_components_dict(data):
...
 
 
 @dbutils.redis_error_handler
-def put_component(component_id):
+def get_component_v3(component_id, state_details=False, config_details=False):
+    """Used by the GET /components/{component_id} API operation"""
+    LOGGER.debug("GET /components/id invoked get_component")
+    if component_id not in DB:
+        return connexion.problem(
+            status=404, title="Component could not found.",
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/components.py:421]:
```suggestion
            status=404, title="Component not found.",
```

File: src/server/cray/cfs/api/controllers/components.py [hunk 2]
```diff
@@ -218,41 +297,156 @@ def patch_v2_components_dict(data):
...
+
+
+def patch_v3_components_list(data):
+    try:
+        components = []
+        for component_data in data:
+            component_id = component_data['id']
+            if component_id not in DB:
+                return connexion.problem(
+                    status=404, title="Component could not found.",
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/components.py:342]:
```suggestion
                    status=404, title="Component not found.",
```

File: src/server/cray/cfs/api/controllers/components.py [hunk 3]
```diff
@@ -218,41 +297,156 @@ def patch_v2_components_dict(data):
...
+    return response, 200
+
+
+@dbutils.redis_error_handler
+def get_component_v2(component_id, config_details=False):
     """Used by the GET /components/{component_id} API operation"""
     LOGGER.debug("GET /components/id invoked get_component")
     if component_id not in DB:
         return connexion.problem(
             status=404, title="Component could not found.",
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/components.py:406]:
```suggestion
            status=404, title="Component not found.",
```

File: src/server/cray/cfs/api/controllers/components.py [hunk 4]
```diff
@@ -261,12 +455,33 @@ def put_component(component_id):
...
 
 
 @dbutils.redis_error_handler
-def patch_component(component_id):
+def patch_component_v2(component_id):
+    """Used by the PATCH /components/{component_id} API operation"""
+    LOGGER.debug("PATCH /components/id invoked patch_component")
+    if component_id not in DB:
+        return connexion.problem(
+            status=404, title="Component could not found.",
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/components.py:469]:
```suggestion
            status=404, title="Component not found.",
```

File: src/server/cray/cfs/api/controllers/components.py [hunk 5]
```diff
@@ -284,7 +499,18 @@ def patch_component(component_id):
 
 
 @dbutils.redis_error_handler
-def delete_component(component_id):
+def delete_component_v2(component_id):
+    """Used by the DELETE /components/{component_id} API operation"""
+    LOGGER.debug("DELETE /components/id invoked delete_component")
+    if component_id not in DB:
+        return connexion.problem(
+            status=404, title="Component could not found.",
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/components.py:507]:
```suggestion
            status=404, title="Component not found.",
```

File: src/server/cray/cfs/api/controllers/components.py [hunk 6]
```diff
@@ -261,12 +455,33 @@ def put_component(component_id):
...
+    data = _set_auto_fields(data)
+    response_data = DB.patch(component_id, data, _update_handler)
+    return convert_component_to_v2(response_data), 200
+
+
+@dbutils.redis_error_handler
+def patch_component_v3(component_id):
     """Used by the PATCH /components/{component_id} API operation"""
     LOGGER.debug("PATCH /components/id invoked patch_component")
     if component_id not in DB:
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/components.py:487]: The "could not found" typo is also a couple lines down, just beyond where github will let me make a suggestion

File: src/server/cray/cfs/api/controllers/components.py [hunk 7]
```diff
@@ -284,7 +499,18 @@ def patch_component(component_id):
...
+            status=404, title="Component could not found.",
+            detail="Component {} could not be found".format(component_id))
+    return DB.delete(component_id), 204
+
+
+@dbutils.redis_error_handler
+def delete_component_v3(component_id):
     """Used by the DELETE /components/{component_id} API operation"""
     LOGGER.debug("DELETE /components/id invoked delete_component")
     if component_id not in DB:
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/components.py:516]: The "could not found" typo is also a couple lines down, just beyond where github will let me make a suggestion

File: src/server/cray/cfs/api/controllers/configurations.py
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/configurations.py:1]: Current year should be added to copyright header

File: src/server/cray/cfs/api/controllers/configurations.py [hunk 8]
```diff
@@ -23,47 +23,103 @@
...
-def get_configurations(in_use=None):
+def get_configurations_v2(in_use=None):
     """Used by the GET /configurations API operation"""
     LOGGER.debug("GET /configurations invoked get_configurations")
-    response = _get_configurations_data(in_use=in_use)
+    configurations_data, next_page_exists = _get_configurations_data(in_use=in_use)
+    if next_page_exists:
+        return connexion.problem(
+            status=400, title="The response size is too large",
+            detail="The response size exceeds the default_page_size.  Use the v3 api to page through the results.")
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/configurations.py:50]:
```suggestion
            detail="The response size exceeds the default_page_size.  Use the v3 API to page through the results.")
```

File: src/server/cray/cfs/api/controllers/sessions.py [hunk 9]
```diff
@@ -159,12 +224,34 @@ def delete_session(session_name):  # noqa: E501
...
+
+    :param session_name: Config Framework Session name
+    :type session_name: str
+
+    :rtype: None
+    """
+    LOGGER.debug("DELETE /v3/sessions/id invoked delete_session")
+    if session_name not in DB:
+        return connexion.problem(
+            status=404, title="Session could not found.",
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/sessions.py:245]:
```suggestion
            status=404, title="Session not found.",
```

File: src/server/cray/cfs/api/controllers/sessions.py [hunk 10]
```diff
@@ -231,11 +369,32 @@ def get_session(session_name):  # noqa: E501
         return connexion.problem(
             status=404, title="Session could not found.",
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/sessions.py:370]:
```suggestion
            status=404, title="Session not found.",
```

File: src/server/cray/cfs/api/controllers/sessions.py [hunk 11]
```diff
@@ -231,11 +369,32 @@ def get_session(session_name):  # noqa: E501
...
+
+    :param session_name: Config Framework Session name
+    :type session_name: str
+
+    :rtype: V3Session
+    """
+    LOGGER.debug("GET /v3/sessions/id invoked get_session")
+    if session_name not in DB:
+        return connexion.problem(
+            status=404, title="Session could not found.",
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/sessions.py:389]:
```suggestion
            status=404, title="Session not found.",
```

File: src/server/cray/cfs/api/controllers/sessions.py [hunk 12]
```diff
@@ -254,12 +413,52 @@ def get_sessions(age=None, min_age=None, max_age=None, status=None, name_contain
...
                 status=400, title="Error parsing the tags provided.",
                 detail=str(err))
-    return _get_filtered_sessions(age, min_age, max_age, status,
-                                  name_contains, succeeded, tag_list), 200
+    sessions_data, next_page_exists = _get_filtered_sessions(age, min_age, max_age, status,
+                                                             name_contains, succeeded, tag_list)
+    if next_page_exists:
+        return connexion.problem(
+            status=400, title="The response size is too large",
+            detail="The response size exceeds the default_page_size.  Use the v3 api to page through the results.")
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/sessions.py:421]:
```suggestion
            detail="The response size exceeds the default_page_size.  Use the v3 API to page through the results.")
```

File: src/server/cray/cfs/api/controllers/sessions.py [hunk 13]
```diff
@@ -281,8 +480,36 @@ def patch_session(session_name):
         return connexion.problem(
             status=404, title="Session could not found.",
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/sessions.py:481]:
```suggestion
            status=404, title="Session not found.",
```

File: src/server/cray/cfs/api/controllers/sessions.py [hunk 14]
```diff
@@ -281,8 +480,36 @@ def patch_session(session_name):
...
+            if key != 'status':
+                raise Exception('Only status can be updated after session creation')
+    except Exception as err:
+        return connexion.problem(
+            status=400, title="Bad Request",
+            detail=str(err))
+
+    if session_name not in DB:
+        return connexion.problem(
+            status=404, title="Session could not found.",
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/sessions.py:509]:
```suggestion
            status=404, title="Session not found.",
```

File: src/server/cray/cfs/api/controllers/versions.py
Comment (mharding-hpe) [inline src/server/cray/cfs/api/controllers/versions.py:1]: Copyright header should be updated to include current year

File: src/server/cray/cfs/api/dbutils.py [hunk 15]
```diff
@@ -89,34 +91,96 @@ def get(self, key):
...
+    def get_keys(self):
+        keys = set()
+        for key in self.client.scan_iter():
+            keys.add(key)
+        # Sorting the keys guarantees a consistent order when paging
+        sorted_keys = sorted(list(keys))
+        return sorted_keys
 
     def put(self, key, new_data):
         """Put data in to the database, replacing any old data."""
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/dbutils.py:144]:
```suggestion
        """Put data into the database, replacing any old data."""
```

File: src/server/cray/cfs/api/dbutils.py [hunk 16]
```diff
@@ -89,34 +91,96 @@ def get(self, key):
...
+                data = self._update(data, patch)
+                if update_handler:
+                    data = update_handler(data)
+                data_str = json.dumps(data)
+                self.client.set(key, data_str)
+                patched_id_list.append(key)
+        return patched_id_list
 
     def _update(self, data, new_data):
         """Recursively patches json to allow sub-fields to be patched."""
```
Comment (mharding-hpe) [inline src/server/cray/cfs/api/dbutils.py:186]:
```suggestion
        """Recursively patches JSON to allow sub-fields to be patched."""
```

**Review by mharding-hpe — APPROVED** [review]
Mostly nits -- looks good!

**Pull request merged**